### PR TITLE
Intrepid2 project fields

### PR DESCRIPTION
**Pull request by mperego**
@trilinos/intrepid2

## Motivation
This PR adds the function `projectField` to project (in `L2` norm) a FE field into a compatible FE space. 
This PR also enables projection tools for `HGRAD_C2` bases, which use the same ordering as `shards` topologies. This required the addition of the missing `HGRAD_LINE_C2_FEM` basis.
The work in this PR is needed by work in #11364. 

Also performed some modifications to avoid out-of-bounds compiler warnings.

Related to  #11364.

## Testing
Added tests both for the new `HGRAD_LINE_C2_FEM` basis, and for the new `projectField` function.

### CODE REVIEW

**Review by rppawlo — APPROVED** [review]
Thanks @mperego !

**Review by CamelliaDPG — APPROVED** [review]
@mperego, looks great!  I called out a few crufty things in the comments, and one set of mesh-generation functions that probably should get replaced with CellGeometry at some point, but I think none of it needs to block the PR.  Up to you if you want to fix the comments before merge.

I look forward to adopting the L^2 projection in Camellia; there are lots of L^2 projections there, and I'm sure your kernel here is much more efficient than mine, and also more performance portable.

File: packages/intrepid2/unit-test/Discretization/Basis/HGRAD_LINE_C2_FEM/CMakeLists.txt
Comment (mperego) [reply packages/intrepid2/unit-test/Discretization/Basis/HGRAD_LINE_C2_FEM/CMakeLists.txt:1]: Yeah, Kyungjoo did that last year (or maybe the year before) when he worked on making Intrepid2 UVM free.

File: packages/intrepid2/unit-test/Projection/test_project_fields.hpp
Comment (mperego) [reply packages/intrepid2/unit-test/Projection/test_project_fields.hpp:1]: That's true, however, the lossy case is covered by other projection tests (e.g., the `test_convergence_XXX`) tests.

**Comment by mperego** [comment]
> @mperego, looks great! I called out a few crufty things in the comments, and one set of mesh-generation functions that probably should get replaced with CellGeometry at some point, but I think none of it needs to block the PR. Up to you if you want to fix the comments before merge.
> 
> I look forward to adopting the L^2 projection in Camellia; there are lots of L^2 projections there, and I'm sure your kernel here is much more efficient than mine, and also more performance portable.

Thanks for the detailed review! I'm going ahead and merge this PR and we can address your suggestions later. In particular, we should create a separate PR and change the contact list in all the Intrepid2 files with your and my contact.

**Comment by CamelliaDPG** [comment]
> Thanks for the detailed review! I'm going ahead and merge this PR and we can address your suggestions later. In particular, we should create a separate PR and change the contact list in all the Intrepid2 files with your and my contact.

Sounds good to me!  Thanks.

**Pull request merged**
